src/easyvk.cpp

```diff
@@ -411,87 +373,140 @@ namespace easyvk
     return subgroupProperties.subgroupSize;
   }
 
-  void Device::teardown()
-  {
+  void Device::teardown() {
     vkDestroyDevice(device, nullptr);
   }
 
-  // Create new buffer
-  VkBuffer getNewBuffer(easyvk::Device &_device, uint32_t size, bool deviceAddr)
-  {
-    VkBufferUsageFlags flags = VK_BUFFER_USAGE_STORAGE_BUFFER_BIT;
-    if (deviceAddr)
-      flags |= VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT;
-    VkBuffer newBuffer;
-    vkCheck(vkCreateBuffer(_device.device, new VkBufferCreateInfo{VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO, nullptr, VkBufferCreateFlags{}, size, flags}, nullptr, &newBuffer));
-    return newBuffer;
-  }
 
-  Buffer::Buffer(easyvk::Device &_device, size_t numElements, size_t elementSize) : Buffer::Buffer(_device, {numElements, elementSize, false, false}) {}
+// -------------------------------------------------------------------------------
 
-  Buffer::Buffer(easyvk::Device &_device, BufferParams params) : device(_device),
-                                                                 buffer(getNewBuffer(_device, params.numElements * params.elementSize, params.deviceAddr)),
-                                                                 _numElements(params.numElements),
-                                                                 _elementSize(params.elementSize),
-                                                                 deviceLocal(params.deviceLocal)
-  {
+  Buffer::Buffer(Device &device, size_t size) : device(device), size(size) {
+    // Create device-local buffer
+    VkBufferCreateInfo bufferInfo {
+      .sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+      .size = size,
+      .usage = VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_STORAGE_BUFFER_BIT,
+      .sharingMode = VK_SHARING_MODE_EXCLUSIVE
+    };
+    vkCheck(vkCreateBuffer(device.device, &bufferInfo, nullptr, &buffer));
 
-    VkMemoryPropertyFlagBits flagBits;
-    if (deviceLocal)
-    {
-      flagBits = VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT;
-    }
-    else
-    {
-      flagBits = VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT;
-    }
-    // Allocate and map memory to new buffer
-    auto memId = _device.selectMemory(buffer, flagBits);
+    // Create staging buffer
```

Comment:
The staging buffer should definitely be optional. On integrated GPUs, there is no need to copy a device local buffer to a staging buffer, since the memory is all the same. This change doubles memory usage on GPUs that have less memory to begin with and reduces performance due to unnecessary copies.

Reply:
Yeah, I had forgotten about the device memory constraints on integrated/mobile GPUs. I'm working on a branch right now to make the staging buffer optional, as the extra memory usage is costing us in sync anyways so I'll have to expedite fixing this.